pkg/provisioner/ironic/ironic.go

```diff
@@ -775,6 +775,59 @@ func (p *ironicProvisioner) UpdateHardwareState() (result provisioner.Result, er
 	return result, nil
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	if p.bmcAccess.RAIDInterface() != "" {
+		// Set raid configuration
+		err = setTargetRAIDCfg(p, ironicNode)
+		if err != nil {
+			p.log.Error(err, "failed in setting the target RAID configuration")
+			return
+		}
+		// Build raid clean steps
+		cleanSteps = append(cleanSteps, buildRAIDCleanSteps(p.host.Status.Provisioning.RAID)...)
+	} else if p.host.Status.Provisioning.RAID != nil {
+		return result, fmt.Errorf("the BMC(%v) of the current node does not support RAID", p.bmcAccess.Driver())
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		result.Dirty = !success
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.CleanWait, nodes.Cleaning, nodes.CleanFail:
+		// Do Nothing (go to provisioning state)
```

Comment:
If something goes bad during the manual cleaning then it should be managed (otherwise I guess the host will transit anyhow in the Provisioning state?)

Reply:
Thanks for your review. After do `startManualCleaning` successfully, the state of `BMH` will be change to the `Provisioning`. We check the result of manual cleaning in the `Provisioning`. But setting `BMH`'s state isn't always successful, so we need to check the node's state, If node's state is `CleanWait` `Cleaning` or `CleanFail` we need let `BMH` go to `Provisioning`, and deal it in `Provisioning`.